src/sdk/pynni/nni/gp_tuner/util.py

```diff
@@ -0,0 +1,171 @@
+    """
+
+    # Warm up with random points
+    x_tries = [space.random_sample()
+               for _ in range(int(num_warmup))]
+    ys = f_acq(x_tries, gp=gp, y_max=y_max)
+    x_max = x_tries[ys.argmax()]
+    max_acq = ys.max()
+
+    # Explore the parameter space more throughly
+    x_seeds = [space.random_sample() for _ in range(num_starting_points)]
+
+    bounds_minmax = np.array(
+        [[bound['_value'][0], bound['_value'][-1]] for bound in bounds])
+
+    for x_try in x_seeds:
+        # Find the minimum of minus the acquisition function
+        res = minimize(lambda x: -f_acq(x.reshape(1, -1), gp=gp, y_max=y_max),
+                       x_try.reshape(1, -1),
+                       bounds=bounds_minmax,
+                       method="L-BFGS-B")
+
+        # See if success
+        if not res.success:
+            continue
+
+        # Store it if better than previous minimum(maximum).
+        if max_acq is None or -res.fun[0] >= max_acq:
+            x_max = _match_val_type(res.x, bounds)
+            max_acq = -res.fun[0]
+
+    # Clip output to make sure it lies within the bounds. Due to floating
+    # point technicalities this is not always the case.
+    return np.clip(x_max, bounds_minmax[:, 0], bounds_minmax[:, 1])
+
+
+class UtilityFunction():
+    """
+    An object to compute the acquisition functions.
+    """
+
+    def __init__(self, kind, kappa, xi):
+        """
+        If UCB is to be used, a constant kappa is needed.
+        """
+        self.kappa = kappa
+
+        self.xi = xi
+
+        if kind not in ['ucb', 'ei', 'poi']:
+            err = "The utility function " \
+                "{} has not been implemented, " \
+                "please choose one of ucb, ei, or poi.".format(kind)
+            raise NotImplementedError(err)
+        self.kind = kind
+
+    def utility(self, x, gp, y_max):
+        '''return utility function'''
+        if self.kind == 'ucb':
+            return self._ucb(x, gp, self.kappa)
+        if self.kind == 'ei':
+            return self._ei(x, gp, y_max, self.xi)
+        if self.kind == 'poi':
+            return self._poi(x, gp, y_max, self.xi)
+        return None
+
+    @staticmethod
+    def _ucb(x, gp, kappa):
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            mean, std = gp.predict(x, return_std=True)
+
+        return mean + kappa * std
+
+    @staticmethod
+    def _ei(x, gp, y_max, xi):
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            mean, std = gp.predict(x, return_std=True)
+
+        z = (mean - y_max - xi)/std
+        return (mean - y_max - xi) * norm.cdf(z) + std * norm.pdf(z)
+
+    @staticmethod
+    def _poi(x, gp, y_max, xi):
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            mean, std = gp.predict(x, return_std=True)
+
+        z = (mean - y_max - xi)/std
```

Comment:
does any possible std == 0.0?

Reply:
Tried to predict a known configuration with GP Regressor, the std is not zero.